src/topology/sober.lean

```diff
@@ -0,0 +1,282 @@
+/-
+Copyright (c) 2021 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import topology.separation
+import topology.continuous_function.basic
+
+/-!
+# Sober spaces
+
+A sober space is a topological space where every irreducible closed subset has a generic point.
+Note that we do not require the generic point to be unique.
+
+## Main definition
+
+* `specializes` : `specializes x y` (`x ⤳ y`) means that `x` specializes to `y`, i.e.
+  `y` is in the closure of `x`.
+* `specialization_preorder` : specialization gives a preorder on the topological space.
+* `specialization_order` : specialization gives a preorder on the topological space.
+* `is_generic_point` : `x` is the generic point of `S` if `S` is the closure of `x`.
+* `sober_space` : A space is sober if every irreducible closed subset has a generic point.
+
+-/
+
+variables {α β : Type*} [topological_space α] [topological_space β]
+
+section specialize_order
+
+/-- `x` specializes to `y` if `y` is in the closure of `x`. The notation used is `x ⤳ y`. -/
+def specializes (x y : α) : Prop := y ∈ closure ({x} : set α)
+
+infix ` ⤳ `:300 := specializes
+
+lemma specializes_def (x y : α) : x ⤳ y ↔ y ∈ closure ({x} : set α) := iff.rfl
+
+lemma specializes_iff_closure_subset {x y : α} :
+  x ⤳ y ↔ closure ({y} : set α) ⊆ closure ({x} : set α) :=
+is_closed_closure.mem_iff_closure_subset
+
+lemma specializes_refl (x : α) : x ⤳ x := subset_closure (set.mem_singleton x)
+
+lemma specializes.trans {x y z : α} : x ⤳ y → y ⤳ z → x ⤳ z :=
+by { simp_rw specializes_iff_closure_subset, exact λ a b, b.trans a }
+
+lemma specializes_iff_forall_closed {x y : α} :
+  x ⤳ y ↔ ∀ (Z : set α) (h : is_closed Z), x ∈ Z → y ∈ Z :=
+begin
+  split,
+  { intros h Z hZ,
+    rw [hZ.mem_iff_closure_subset, hZ.mem_iff_closure_subset],
+    exact (specializes_iff_closure_subset.mp h).trans },
+  { intro h, exact h _ is_closed_closure (subset_closure $ set.mem_singleton x) }
+end
+
+lemma specializes_iff_forall_open {x y : α} :
+  x ⤳ y ↔ ∀ (U : set α) (h : is_open U), y ∈ U → x ∈ U :=
+begin
+  rw specializes_iff_forall_closed,
+  exact ⟨λ h U hU, not_imp_not.mp (h _ (is_closed_compl_iff.mpr hU)),
+    λ h U hU, not_imp_not.mp (h _ (is_open_compl_iff.mpr hU))⟩,
+end
+
+lemma indistinguishable_iff_specialization_or (x y : α) :
+  indistinguishable x y ↔ x ⤳ y ∧ y ⤳ x :=
+(indistinguishable_iff_closure x y).trans (and_comm _ _)
+
+lemma indistinguishable.eq [hα : t0_space α] {x y : α} (h : indistinguishable x y) : x = y :=
+not_imp_not.mp ((t0_space_iff_distinguishable _).mp hα x y) h
+
+lemma specializes_antisymm [t0_space α] (x y : α) : x ⤳ y → y ⤳ x → x = y :=
+λ h₁ h₂, ((indistinguishable_iff_specialization_or _ _).mpr ⟨h₁, h₂⟩).eq
+
+lemma specializes.map {x y : α} (h : x ⤳ y) (f : C(α, β)) : f x ⤳ f y :=
```

Comment:
Should this maybe be `continuous_map.map_specializes`? And then `specializes.map` can use unbundled continuous functions? I think it's useful to have both.